closeio_api/__init__.py

```diff
@@ -44,110 +50,196 @@ def __init__(self, base_url, api_key=None, tz_offset=None,
+        return func('get', endpoint+'/', **kwargs)
 
     def post(self, endpoint, data, **kwargs):
+        """Send (sync client) or prepare (async client) a POST request
+        to a given endpoint, for example:
+
+        >>> api.post('lead', {'name': 'Brand New Lead'})
+        {
+            'name': 'Brand New Lead'
+            # ... rest of the response omitted for brevity
+        }
+        """
         kwargs.update({'data': data})
-        return self.dispatch('post', endpoint+'/', **kwargs)
+        func = self._prepare_request if self.async else self._dispatch
+        return func('post', endpoint+'/', **kwargs)
 
     def put(self, endpoint, data, **kwargs):
+        """Send (sync client) or prepare (async client) a PUT request to
+        a given endpoint, for example:
+
+        >>> api.put('lead/SOME_LEAD_ID', {'name': 'New Name'})
+        {
+            'name': 'New Name'
+            # ... rest of the response omitted for brevity
+        }
+        """
         kwargs.update({'data': data})
-        return self.dispatch('put', endpoint+'/', **kwargs)
+        func = self._prepare_request if self.async else self._dispatch
+        return func('put', endpoint+'/', **kwargs)
 
     def delete(self, endpoint, **kwargs):
-        return self.dispatch('delete', endpoint+'/', **kwargs)
+        """Send (sync client) or prepare (async client) a DELETE request
+        to a given endpoint, for example:
+
+        >>> api.delete('lead/SOME_LEAD_ID')
+        {}
+        """
+        func = self._prepare_request if self.async else self._dispatch
+        return func('delete', endpoint+'/', **kwargs)
 
     # Only for async requests
     def map(self, reqs, max_retries=None):
+        """Execute a batch of asynchronous requests concurrently. For
+        example:
+
+        >>> reqs = []
+        >>> reqs.append(api.post('lead', {'name': 'New Lead'}))
+        >>> reqs.append(api.post('lead', {'name': 'Another New Lead'}))
+        >>> api.map(reqs)
+        [
+            # list of dicts containing successful API responses and APIError
+            # objects for failed requests.
+        ]
+        """
+        if not self.async:
+            raise NotImplementedError('map can only be used in an async Client')
+
         if max_retries is None:
             max_retries = self.max_retries
+
         # TODO
         # There is no good way of catching or dealing with exceptions that are
         # raised during the request sending process when using map or imap.
         # When this issue is closed:
         # https://github.com/kennethreitz/grequests/pull/15
         # modify this method to repeat only the requests that failed because of
         # connection errors
-        if self.async:
-            import grequests
-            responses = [(
+
+        responses = [(
+            response.json() if response.ok else APIError(response)
+        ) for response in self.requests.map(reqs)]
```

Comment:
The pool needs to have a (safe) size to prevent firing too many concurrent API requests.